torchtext/datasets/amazonreviewpolarity.py

```diff
@@ -34,14 +34,18 @@
 @_add_docstring_header(num_lines=NUM_LINES, num_classes=2)
 @_create_dataset_directory(dataset_name=DATASET_NAME)
 @_wrap_split_argument(("train", "test"))
-def AmazonReviewPolarity(root: str, split: Union[Tuple[str], str]):
+def AmazonReviewPolarity(root: str, split: Union[Tuple[str], str], validate_hash: bool = True):
```

Comment:
You might want to be careful on adding a public constructor arguments for the sake of the tests (though this looks like an argument with valid usecases).

What do you anticipate the future extension of this dataset? If new options related to dataset are added, which will be used more often, then having this argument before it might not look elegant in the end. At the same time, if all the dataset is going to have the same argument, it looks consistent.

If this is intended only for testing, it is worth considering if it is possible to inject the hash in a different way. For example you can use `patch` to overwrite the value of `MD5` for the duration of the test.

https://stackoverflow.com/a/51567124

Reply:
I think patching approach make sense to me. I would be careful about exposing `validate_hash` to users.

Reply:
I think I ended up taking this approach because we used a `validate_hash` arg for testing the SST2 dataset (https://github.com/pytorch/text/pull/1438). I am trying to figure out how to mock the value of `MD5` and the `hash_dict` being passed into `on_disk_cache` but can't seem to figure out a good way to do this. I was thinking of following the suggestions in [this thread](https://stackoverflow.com/questions/18191275/using-pythons-mock-patch-object-to-change-the-return-value-of-a-method-called-w) but I'm not sure if we can mock out the `on_disk_cache` like this (especially since we call it twice and pass in different args to it each time). Do you guys have any tips or code pointers that might help?

cc @ejguan  @mthrok @parmeet

Reply:
> how to mock the value of `MD5` and the `hash_dict` being passed into `on_disk_cache`

IIUC, once the tar file is created, then you can compute the hash value of the tar, and patch the value of `torchtext.datasets.amazonreviewpolarity.MD5`. 
Mocking the `on_disk_cache` is an alternative way to do so, but IMO, it's less flexible, because it requires precise understanding of how and when the function is called, which can be changed anytime in the future. 

Something like the following should work.  the patch command in `setUpClass` then undo the patch in `tearDownClass`.
https://docs.python.org/3/library/unittest.mock.html#patch-methods-start-and-stop

```python
def setUpClass(cls):
    super().setUpClass()
    cls.root_dir = cls.get_base_temp_dir()
    cls.samples = get_mock_dataset(cls.root_dir)
    val = compute_hash(<the archive file>)
    cls.patcher = patch('torchtext.datasets.amazonreviewpolarity.MD5', val)
    cls.patcher.start()

def tearDownClass(cls):
    super().tearDownClass()
    cls.patcher.stop()
```

Reply:
With this argument, users can determine if they want to check the hash of cached files. You guys can decide if users would turn this off to skip hash check.
Note: If the concern comes from API forward compatibility, it can be switched to a keyword argument to move it to the the end.

Reply:
@mthrok thanks for the example. Let me try it with your approach to see if I can get this working.

@ejguan thanks for the input. I think we were only adding the `validate_hash` arg to make it easier for testing since we don't expect the user to modify the downloaded files so they should have no need to skip the hash check. If we can get the patching to work, then I will get rid of the `validate_hash` arg.

Reply:
I think the patch should be

```python
with patch("torchdata.datapipes.iter.util.cacheholder._hash_check", return_value=True):
   ...
```

Reply:
> I think a slightly more to-the-spec patch would be
> 
> ```python
> # Who cares what the MD5 is, we care about the behavior if hashes match
> with patch("torchdata.datapipes.iter.util.cacheholder._hash_check", return_value=True):
>    ...
> ```

Actually that does seem like the better approach, thanks @erip!

Reply:
@erip your suggestion for patching "torchdata.datapipes.iter.util.cacheholder._hash_check" to return `True` worked really well here! Would you mind reviewing this PR one more time before I merge it in?

Reply:
I think this looks good!